docs/architecture/ADR-003-borrow-assets.md

```diff
@@ -5,106 +5,155 @@
+For the purposes of borrowing and repaying assets, as well as marking uTokens as collateral, the `umee/x/leverage` module does not mint or burn tokens. It stores borrower open positions and collateral settings, and uses the `x/bank` module to perform all necessary balance checks and token transfers. User collateral (uTokens) are deposited in `x/leverage` module and withdrawn back to the user `x/bank` account balance when the user disables uTokens as a collateral.
 
-To prevent users from transferring uTokens when doing so would lower their borrow limit below their current total borrowed amount, uTokens that are enabled as collateral are stored in the `x/leverage` module account, and moved back to the user's wallet when they are disabled as collateral.
-As long as it would not reduce a user's borrow limit below their borrowed value, `MsgWithdrawAsset` can withdraw base assets in exchange uTokens used as collateral, even when they are stored in the module account.
+During every operation which involves borrow position or collateral we check that the _borrow limit_ rule holds.
 
-Because borrow limits weight the value of different token denominations together, the calculation will require the use of `x/oracle`.
+`x/oracle` module is used to provide exchange rates of tokens to calculate borrow limit.
 
-### Basic Message Types
+### API
 
-To implement the borrow/repay functionality of the Asset Facility, the three common message types will be:
+To implement the borrow/repay functionality of the Asset Facility, the three message types are defined:
 
 ```go
-// MsgSetCollateral - a borrower enables or disables a specific utoken type in their wallet to be used as collateral
+// MsgSetCollateral - a borrower enables or disables a specific uToken type in their wallet to be used as collateral
 type MsgSetCollateral struct {
   Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
   Denom    string         `json:"denom" yaml:"denom"`
   Enable   bool           `json:"enable" yaml:"enable"`
 }
+
 // MsgBorrowAsset - a user wishes to borrow assets of an allowed type
 type MsgBorrowAsset struct {
   Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
+  // not a uToken
   Amount   sdk.Coin       `json:"amount" yaml:"amount"`
 }
+
 // MsgRepayAsset - a user wishes to repay assets of a borrowed type
 type MsgRepayAsset struct {
   Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
   Amount   sdk.Coin       `json:"amount" yaml:"amount"`
 }
+
+// MsgSetCollateralMaxUtilization is going to be used as CosmosSDK 0.46 gov message.
+type MsgSetCollateralMaxUtilization {
+  // must be x/gov
+  string signer = 1;
+  string token = 2;
+  // a number between 0 and 10'000 - ratio in basis points.
+  uint32 max_utilization = 3;
+}
```

Comment:
```suggestion
```
Per-token parameters should be fields in the `Token` struct (stored in registry, updated as a whole by governance)

Reply:
OK, added to token registry. BTW - we will need to update the registry as well. I'm adding a task for that.